src/main/java/org/opensearch/security/privileges/ResourceAccessEvaluator.java

```diff
@@ -0,0 +1,212 @@
+                    ? "Publicly shared resource"
+                    : "User " + user.getName() + " is the owner of the resource";
+                log.debug("{} {}, granting access.", message, req.id());
+                latch.countDown();
+                return;
+            }
+
+            // Check whether user or their roles match any access-levels this resource is shared at
+            Set<String> accessLevels = new HashSet<>();
+            accessLevels.addAll(document.fetchAccessLevels(Recipient.USERS, Set.of(user.getName())));
+            accessLevels.addAll(document.fetchAccessLevels(Recipient.ROLES, userRoles));
+            accessLevels.addAll(document.fetchAccessLevels(Recipient.BACKEND_ROLES, userBackendRoles));
+
+            // if no access-levels match, then action is not allowed
+            if (accessLevels.isEmpty()) {
+                presponse.allowed = false;
+                log.debug("Resource {} is not shared with user {}", req.id(), user.getName());
+                shouldMarkAsComplete.set(true);
+                latch.countDown();
+                return;
+            }
+
+            // Expand access-levels and check if any actions match the action supplied
+            if (context.getActionPrivileges() instanceof RoleBasedActionPrivileges roleBasedActionPrivileges) {
+                Set<String> actions = roleBasedActionPrivileges.flattenedActionGroups().resolve(accessLevels);
+                // a matcher to test against all patterns in `actions`
+                WildcardMatcher matcher = WildcardMatcher.from(actions, true);
+                if (matcher.test(action)) {
+                    presponse.allowed = true;
+                    log.debug("Resource {} is shared with user {}, granting access.", req.id(), user.getName());
+                } else {
+                    // TODO check why following addition doesn't reflect in the final response message and find an alternative
+                    presponse.getMissingPrivileges().add(action);
+                    log.debug("User {} has no {} privileges for {}", user.getName(), action, req.id());
+                }
+                latch.countDown();
+            } else {
+                // NOTE we don't yet support Plugins to access resources
+                presponse.allowed = false;
+                log.debug(
+                    "Plugin access to resources is currently not supported. Plugin {} is not authorized to access resource {}.",
+                    user.getName(),
+                    req.id()
+                );
+                latch.countDown();
+            }
+            shouldMarkAsComplete.set(true);
+
+        }, e -> {
+            presponse.allowed = false;
+            log.debug("Something went wrong while evaluating resource {}. Marking request as unauthorized.", req.id());
+            shouldMarkAsComplete.set(true);
+            latch.countDown();
+        }));
+        try {
+            latch.await();
```

Comment:
Unfortunately, blocking on index operations here bears the risk of spurious deadlocks due to thread pool exhaustion ...

Reply:
But that's a risk of dead locks; i am not sure that this can be classified as optimization issue.

Reply:
oh I see, it would be a netty thread for HTTP Request handling, but at the transport level we can specify executor for a TransportAction. For example: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/explain/TransportExplainAction.java#L203-L209